src/main/host/syscall/unistd.rs

```diff
@@ -93,214 +91,170 @@ pub fn dup_helper(
     };
 
     // return the new fd
-    SyscallReturn::Success(new_fd).into()
+    Ok(new_fd.into())
 }
 
-pub fn read(sys: &mut c::SysCallHandler, args: &c::SysCallArgs) -> c::SysCallReturn {
-    let fd = unsafe { args.args[0].as_i64 } as libc::c_int;
-    let buf_ptr = unsafe { args.args[1].as_ptr };
-    let buf_size = unsafe { args.args[2].as_u64 } as libc::size_t;
+pub fn read(sys: &mut c::SysCallHandler, args: &SysCallArgs) -> SyscallResult {
+    let fd = libc::c_int::from(args.get(0));
+    let buf_ptr = PluginPtr::from(args.get(1));
+    let buf_size = libc::size_t::from(args.get(2));
     let offset = 0;
 
     // get the descriptor, or return early if it doesn't exist
-    let desc = match syscall::get_descriptor(fd, sys.process) {
-        Ok(d) => unsafe { &mut *d },
-        Err(errno) => return SyscallReturn::Error(errno).into(),
-    };
+    let desc = unsafe { &mut *syscall::get_descriptor(fd, sys.process)? };
 
     match desc {
         CompatDescriptor::New(desc) => read_helper(sys, fd, desc, buf_ptr, buf_size, offset),
         // if it's a legacy descriptor, use the C syscall handler instead
         CompatDescriptor::Legacy(_) => unsafe {
-            c::syscallhandler_read(sys as *mut c::SysCallHandler, args as *const c::SysCallArgs)
+            c::syscallhandler_read(sys as *mut c::SysCallHandler, args as *const SysCallArgs).into()
         },
```

Comment:
> it would be nice to have some tests that try different values to make sure we don't accidentally lose information during the conversion, and that we end up with what we started with. 

Done

> Maybe it would also be a good idea to add a test that checks the size of c::SysCallReturn

I think this is covered by the conversion tests; they'll fail to compile due to incomplete initializers if a new field is added.